README.MD
# More Sponges Mod

#### What this mod adds?

- Craftable sponges 😲
- large and extra large sponges 🧺
- Normal, Large and Extra large lava 🔥 Sponges


changelog.txt
Build: 1.15.2-31.1.76 - Wed May 06 20:41:52 GMT 2020
	jmansfield:
		Fix ModListScreen escape key not matching done button behavior. Closes #6672

=========
Build: 1.15.2-31.1.75 - Wed May 06 11:31:06 GMT 2020
	David Quintana:
		Patch the bucket logic to not hardcode some checks to WATER. Fixes not being able to place custom fluids into modded fluid-loggable blocks that accept fluids other than water.

=========
Build: 1.15.2-31.1.74 - Sun May 03 23:13:58 GMT 2020
	David Quintana:
		Avoid discarding the bottom 4 bits of the lightmap coords. It turns out vanilla relies on those bits. Fixes some blocks being darker than they should be in certain circumstances.

=========
Build: 1.15.2-31.1.73 - Sun May 03 20:48:25 GMT 2020
	jacob101607:
		Fix issue with providing a null ItemStack to onPlayerDestroyItem (#6633)

=========
Build: 1.15.2-31.1.72 - Sun May 03 20:40:43 GMT 2020
	Bward7864:
		Allow overriding of tool level requirements when harvesting a block, via the HarvestCheck event.  (#6632)

=========
Build: 1.15.2-31.1.71 - Sun May 03 20:34:34 GMT 2020
	richard:
		Fixed Hoppers not inserting fully into custom contains with slots that can hold more then max stack (#6631)

=========
Build: 1.15.2-31.1.70 - Sun May 03 20:27:39 GMT 2020
	darklime:
		Change recipe condition log level to debug. (#6621)

	darklime:
		Fix vanilla loot table resource leak. (#6622)

	LordMonoxide:
		Fixed ChunkDataEvent.Save being fired with a null world if chunk isn't fully generated when unloaded. (#6628)

=========
Build: 1.15.2-31.1.67 - Sun May 03 20:21:40 GMT 2020
	41904979+CAS-ual-TY:
		Fix milk buckets removing potion effects without filters. (#6609)

=========
Build: 1.15.2-31.1.66 - Sun May 03 20:15:40 GMT 2020
	dion:
		Added missing localization for swim speed, and reach distance attributes (#6585)

=========
Build: 1.15.2-31.1.65 - Sun May 03 20:09:37 GMT 2020
	williewillus:
		Allow colors of tooltip to be specified in GuiUtils.drawHoveringText tooltip methods (#6579)

=========
Build: 1.15.2-31.1.64 - Sun May 03 19:56:55 GMT 2020
	ichttt:
		Fix config comments referencing disallowed enum values. (#6542)

=========
Build: 1.15.2-31.1.63 - Tue Apr 28 17:23:47 GMT 2020
	David Quintana:
		Remove redundant call to modifyLoot. Closes #6657

=========
Build: 1.15.2-31.1.62 - Tue Apr 28 04:02:06 GMT 2020
	tterrag:
		Fix various issues with the loading screen and datagen
		
		Fix NPE on datagen startup
		Fix early loading GUI hanging and never terminating after datagen
		Add loading GUI messages for datagen

=========
Build: 1.15.2-31.1.61 - Mon Apr 27 18:06:44 GMT 2020
	weiland:
		Fix BackgroundScanHandler crashing on servers (#6660)

=========
Build: 1.15.2-31.1.60 - Mon Apr 27 13:55:33 GMT 2020
	cpw:
		Remove silly icon override. Squirrels are fun y'all..
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.59 - Mon Apr 27 05:10:06 GMT 2020
	tterrag:
		Fix VariantList calling deprecated bake method

	tterrag:
		Rename window icon image to avoid conflict, fix not using vanilla icon

=========
Build: 1.15.2-31.1.57 - Mon Apr 27 03:06:13 GMT 2020
	lex:
		Fix Automatic Event Subscriber not detecting mod id defaults, and fix test mods.

	lex:
		Fix resource packs not being sorted properly. Closes #6287

=========
Build: 1.15.2-31.1.55 - Mon Apr 27 01:42:42 GMT 2020
	roguelogix:
		move handOffWindow down near start()

	roguelogix:
		Move ClientVisualization to secondary thread

	cpw:
		Handle message display a bit better, try to make the UI tick a bit better.. Still a problem with
		a modelbake right at the end, happening on the window flip.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

	cpw:
		Use forge logo as short term window Icon, fix Buffer cast problems for new JDK stupid.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.51 - Sun Apr 26 22:25:05 GMT 2020
	cpw:
		merge early startup into main window if used. It seems to work well.. Also ticks window on main thread, so mac compatible now.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.50 - Sat Apr 25 21:55:22 GMT 2020
	cpw:
		Update jenkinsfile and gradle for new versions of things..
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.49 - Sat Apr 25 02:19:14 GMT 2020
	lex:
		Fix potential NPEs in RegistryObject.orElseThrow/isPresent/ifPresent Closes #6647

=========
Build: 1.15.2-31.1.48 - Thu Apr 23 12:57:42 GMT 2020
	David Quintana:
		Add tickRate setter to the ForgeFlowingFluid Properties builder.

=========
Build: 1.15.2-31.1.47 - Mon Apr 20 01:37:16 GMT 2020
	cpw:
		Fix recursion mess in PlayerAdvancement loading, by using an alternative based on a toposorted list of all advancements.
		It seems to be fully equivalent in testing, but there may be corner cases, so there is a config to disable, though disabling
		may break servers, especially those chunkloading fake players.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.15.2-31.1.46 - Wed Apr 15 17:10:05 GMT 2020
	lex:
		Revert defualt jvm args.

=========
Build: 1.15.2-31.1.45 - Wed Apr 15 06:54:47 GMT 2020
	lex:
		Fixed dev time compilation using J9+ producing incorrect bytecode.

=========
Build: 1.15.2-31.1.44 - Sun Apr 12 19:27:19 GMT 2020
	David Quintana:
		Add "type" field to global loot modifier jsons. (#6554)
		
		* Add "type" field to global loot modifier jsons and fix tests.
		To avoid binary breaking, "type" defaults to be the same as the filename.
		* Fix license headers.

=========
Build: 1.15.2-31.1.43 - Sat Apr 11 06:03:58 GMT 2020
	tterrag:
		Fix child annotations not inheriting member name, cleanup

=========
Build: 1.15.2-31.1.42 - Sat Apr 11 05:20:15 GMT 2020
	tterrag:
		Fix element type of scanned child annotations

=========
Build: 1.15.2-31.1.41 - Thu Apr 09 00:19:35 GMT 2020
	28677702+patrick-aoc:
		Fixed PlantType not extending IExtensibleEnum, Closes #6286

=========
Build: 1.15.2-31.1.40 - Thu Apr 09 00:12:15 GMT 2020

src/main/java/io/github/tomat1337/morespongemod/ExampleMod.java
package io.github.tomat1337.morespongemod;

import net.minecraft.block.Block;
import net.minecraft.block.Blocks;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.InterModComms;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
import net.minecraftforge.fml.event.lifecycle.InterModEnqueueEvent;
import net.minecraftforge.fml.event.lifecycle.InterModProcessEvent;
import net.minecraftforge.fml.event.server.FMLServerStartingEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.stream.Collectors;

// The value here should match an entry in the META-INF/mods.toml file
@Mod(ModInfo.MOD_ID)
public class ExampleMod
{
    // Directly reference a log4j logger.
    private static final Logger LOGGER = LogManager.getLogger();

    public ExampleMod() {
        // Register the setup method for modloading
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::setup);
        // Register the enqueueIMC method for modloading
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::enqueueIMC);
        // Register the processIMC method for modloading
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::processIMC);
        // Register the doClientStuff method for modloading
        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::doClientStuff);

        // Register ourselves for server and other game events we are interested in
        MinecraftForge.EVENT_BUS.register(this);
    }

    private void setup(final FMLCommonSetupEvent event)
    {
        // some preinit code
        LOGGER.info("HELLO FROM PREINIT");
        LOGGER.info("DIRT BLOCK >> {}", Blocks.DIRT.getRegistryName());
    }

    private void doClientStuff(final FMLClientSetupEvent event) {
        // do something that can only be done on the client
        LOGGER.info("Got game settings {}", event.getMinecraftSupplier().get().gameSettings);
    }

    private void enqueueIMC(final InterModEnqueueEvent event)
    {
        // some example code to dispatch IMC to another mod
        InterModComms.sendTo("examplemod", "helloworld", () -> { LOGGER.info("Hello world from the MDK"); return "Hello world";});
    }

    private void processIMC(final InterModProcessEvent event)
    {
        // some example code to receive and process InterModComms from other mods
        LOGGER.info("Got IMC {}", event.getIMCStream().
                map(m->m.getMessageSupplier().get()).
                collect(Collectors.toList()));
    }
    // You can use SubscribeEvent and let the Event Bus discover methods to call
    @SubscribeEvent
    public void onServerStarting(FMLServerStartingEvent event) {
        // do something when the server starts
        LOGGER.info("HELLO from server starting");
    }

    // You can use EventBusSubscriber to automatically subscribe events on the contained class (this is subscribing to the MOD
    // Event bus for receiving Registry Events)
    @Mod.EventBusSubscriber(bus=Mod.EventBusSubscriber.Bus.MOD)
    public static class RegistryEvents {
        @SubscribeEvent
        public static void onBlocksRegistry(final RegistryEvent.Register<Block> blockRegistryEvent) {
            // register a new block here
            LOGGER.info("HELLO from Register Block");
        }
    }
}


src/main/java/io/github/tomat1337/morespongemod/Blocks/SpongeBlock.java
package io.github.tomat1337.morespongemod.Blocks;

import com.google.common.collect.Lists;
import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.fluid.Fluids;
import net.minecraft.fluid.IFluidState;
import net.minecraft.tags.FluidTags;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.Tuple;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import java.util.Queue;

public class SpongeBlock extends Block {
    private boolean lava;
    private int range;
    private char type;

    public SpongeBlock(Properties properties, boolean islava, int range,char type) {
        super(properties);
        this.lava = islava;
        this.type = type;
        this.range = range;
    }
    public void onBlockAdded(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving) {
        if (oldState.getBlock() != state.getBlock()) {
            this.tryAbsorb(worldIn, pos);
        }
    }

    public void neighborChanged(BlockState state, World worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving) {
        this.tryAbsorb(worldIn, pos);
        super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving);
    }

    protected void tryAbsorb(World worldIn, BlockPos pos) {
        if (this.absorb(worldIn, pos)) {
            //Replace Sponge to Wet Sponge
            if (lava) {
                if (type == 'E')//extra large
                    worldIn.setBlockState(pos, BlocksRegistry.wet_lava_extra_large_sponge.getDefaultState(), 2);
                else if (type == 'L')//large
                    worldIn.setBlockState(pos, BlocksRegistry.wet_lava_large_sponge.getDefaultState(), 2);
                else// normal
                    worldIn.setBlockState(pos, BlocksRegistry.wet_lava_sponge.getDefaultState(), 2);

            } else {
                if (type == 'E')//extra large
                    worldIn.setBlockState(pos, BlocksRegistry.wet_extra_large_sponge.getDefaultState(), 2);
                else if (type == 'L')//large
                    worldIn.setBlockState(pos, BlocksRegistry.wet_large_sponge.getDefaultState(), 2);
            }
            worldIn.playEvent(2001, pos, Block.getStateId(Blocks.WATER.getDefaultState()));
        }

    }

    private boolean absorb(World worldIn, BlockPos pos) {
        Queue<Tuple<BlockPos, Integer>> queue = Lists.newLinkedList();
        queue.add(new Tuple<>(pos, 0));
        int i = 0;
        int size = 6 + range;
        while(!queue.isEmpty()) {
            Tuple<BlockPos, Integer> tuple = queue.poll();
            BlockPos blockpos = tuple.getA();
            int j = tuple.getB();

            for(Direction direction : Direction.values()) {

                BlockPos blockpos1 = blockpos.offset(direction);
                BlockState blockstate = worldIn.getBlockState(blockpos1);
                IFluidState ifluidstate = worldIn.getFluidState(blockpos1);
                Material material = blockstate.getMaterial();

                if (ifluidstate.isTagged(FluidTags.WATER) && lava == false) { // water sponge
                    if (blockstate.getBlock() instanceof IBucketPickupHandler && ((IBucketPickupHandler)blockstate.getBlock()).pickupFluid(worldIn, blockpos1, blockstate) != Fluids.EMPTY) {
                        ++i;
                        if (j < size) { // change 6 for my  custom size value does not change anything
                            queue.add(new Tuple<>(blockpos1, j + 1));
                        }
                    } else if (blockstate.getBlock() instanceof FlowingFluidBlock) {
                        worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 3);
                        ++i;
                        if (j < size) { // change 6 for my  custom size value does not change anything
                            queue.add(new Tuple<>(blockpos1, j + 1));
                        }
                    } else if (material == Material.OCEAN_PLANT || material == Material.SEA_GRASS) {
                        TileEntity tileentity = blockstate.getBlock().hasTileEntity() ? worldIn.getTileEntity(blockpos1) : null;
                        spawnDrops(blockstate, worldIn, blockpos1, tileentity);
                        worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 3);
                        ++i;
                        if (j < size) {
                            queue.add(new Tuple<>(blockpos1, j + 1));
                        }
                    }
                }else if (ifluidstate.isTagged(FluidTags.LAVA) && lava) { // Lava sponge
                    if (blockstate.getBlock() instanceof IBucketPickupHandler && ((IBucketPickupHandler)blockstate.getBlock()).pickupFluid(worldIn, blockpos1, blockstate) != Fluids.EMPTY) {
                        ++i;
                        if (j < size) { // change 6 for my  custom size value does not change anything
                            queue.add(new Tuple<>(blockpos1, j + 1));
                        }
                    } else if (blockstate.getBlock() instanceof FlowingFluidBlock) {
                        worldIn.setBlockState(blockpos1, Blocks.AIR.getDefaultState(), 3);
                        ++i;
                        if (j < size) { // change 6 for my  custom size value does not change anything
                            queue.add(new Tuple<>(blockpos1, j + 1));
                        }
                    }
                }
            }

            if (i > 64 * size) { // dont know how tis cod works but changing this value changes de size off the succ
                break;
            }
        }

        return i > 0;
    }
}


src/main/java/io/github/tomat1337/morespongemod/Blocks/WetSpongeBlock.java
package io.github.tomat1337.morespongemod.Blocks;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.particles.ParticleTypes;
import net.minecraft.util.Direction;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvents;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import java.util.Random;

public class WetSpongeBlock  extends Block {
    private boolean islava;
    private char type;

    protected WetSpongeBlock(Properties properties, boolean islava,char type) {
        super(properties);
        this.islava = islava;
        this.type = type;
    }

    public void onBlockAdded(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving) {
        if (worldIn.getDimension().doesWaterVaporize() && !islava) {//water sponges
            if (type == 'E')//extra large
                worldIn.setBlockState(pos, BlocksRegistry.extra_large_sponge.getDefaultState(), 3);
            else if (type == 'L')//large
                worldIn.setBlockState(pos, BlocksRegistry.large_sponge.getDefaultState(), 3);
            worldIn.playEvent(2009, pos, 0);
            worldIn.playSound((PlayerEntity) null, pos, SoundEvents.BLOCK_FIRE_EXTINGUISH, SoundCategory.BLOCKS, 1.0F, (1.0F + worldIn.getRandom().nextFloat() * 0.2F) * 0.7F);
        }
        if (!worldIn.getDimension().doesWaterVaporize() && islava) {//lava sponges
            if (type == 'E')//extra large
                worldIn.setBlockState(pos, BlocksRegistry.lava_extra_large_sponge.getDefaultState(), 3);
            else if (type == 'L')//large
                worldIn.setBlockState(pos, BlocksRegistry.lava_large_sponge.getDefaultState(), 3);
            else// normal
                worldIn.setBlockState(pos, BlocksRegistry.lava_sponge.getDefaultState(), 3);
            worldIn.playEvent(2009, pos, 0);
            worldIn.playSound((PlayerEntity) null, pos, SoundEvents.AMBIENT_UNDERWATER_ENTER, SoundCategory.BLOCKS, 1.0F, (1.0F + worldIn.getRandom().nextFloat() * 0.2F) * 0.7F);
        }

    }

    @OnlyIn(Dist.CLIENT)
    public void animateTick(BlockState stateIn, World worldIn, BlockPos pos, Random rand) {
        Direction direction = Direction.random(rand);
        if (direction != Direction.UP) {
            BlockPos blockpos = pos.offset(direction);
            BlockState blockstate = worldIn.getBlockState(blockpos);
            if (!stateIn.isSolid() || !blockstate.isSolidSide(worldIn, blockpos, direction.getOpposite())) {
                double d0 = (double)pos.getX();
                double d1 = (double)pos.getY();
                double d2 = (double)pos.getZ();
                if (direction == Direction.DOWN) {
                    d1 = d1 - 0.05D;
                    d0 += rand.nextDouble();
                    d2 += rand.nextDouble();
                } else {
                    d1 = d1 + rand.nextDouble() * 0.8D;
                    if (direction.getAxis() == Direction.Axis.X) {
                        d2 += rand.nextDouble();
                        if (direction == Direction.EAST) {
                            ++d0;
                        } else {
                            d0 += 0.05D;
                        }
                    } else {
                        d0 += rand.nextDouble();
                        if (direction == Direction.SOUTH) {
                            ++d2;
                        } else {
                            d2 += 0.05D;
                        }
                    }
                }
                if(islava) {
                    if (rand.nextInt(5) == 0) {
                        for (int i = 0; i < rand.nextInt(1) + 1; ++i) {
                            worldIn.addParticle(ParticleTypes.LAVA, (double) ((float) pos.getX() + 0.5F), (double) ((float) pos.getY() + 0.5F), (double) ((float) pos.getZ() + 0.5F), (double) (rand.nextFloat() / 2.0F), 5.0E-5D, (double) (rand.nextFloat() / 2.0F));
                            worldIn.addParticle(ParticleTypes.CAMPFIRE_COSY_SMOKE, (double) ((float) pos.getX() + 0.5F), (double) ((float) pos.getY() + 0.5F), (double) ((float) pos.getZ() + 0.5F), (double) (rand.nextFloat() / 2.0F), 5.0E-5D, (double) (rand.nextFloat() / 2.0F));
                        }
                    }
                    worldIn.addParticle(ParticleTypes.LANDING_LAVA, d0, d1, d2, 0.0D, 0.0D, 0.0D);
                }
                else
                    worldIn.addParticle(ParticleTypes.DRIPPING_WATER, d0, d1, d2, 0.0D, 0.0D, 0.0D);
            }
        }
    }
}


src/main/java/io/github/tomat1337/morespongemod/Blocks/BlocksRegistry.java
package io.github.tomat1337.morespongemod.Blocks;

import io.github.tomat1337.morespongemod.ModInfo;
import net.minecraft.block.Block;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraftforge.common.ToolType;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.registries.ObjectHolder;

@ObjectHolder(ModInfo.MOD_ID)
@Mod.EventBusSubscriber(modid = ModInfo.MOD_ID, bus = Mod.EventBusSubscriber.Bus.MOD)
public class BlocksRegistry {
    //Sponge Blocks
    public static Block large_sponge =
            new SpongeBlock(Block.Properties
                    .create(Material.SPONGE)
                    .harvestTool(ToolType.get("sword"))
                    .sound(SoundType.CLOTH)
                    .hardnessAndResistance(0.5f, 0.5f)
                    , false
                    , 7
                    ,'L');
    public static Block extra_large_sponge =
            new SpongeBlock(Block.Properties
                    .create(Material.SPONGE)
                    .harvestTool(ToolType.get("sword"))
                    .sound(SoundType.CLOTH)
                    .hardnessAndResistance(0.5f, 0.5f)
                    , false
                    , 14
                    ,'E');
    public static Block lava_sponge =
            new SpongeBlock(Block.Properties
                    .create(Material.SPONGE)
                    .harvestTool(ToolType.get("sword"))
                    .sound(SoundType.CLOTH)
                    .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    , 1
                    ,'N');
    public static Block lava_large_sponge =
            new SpongeBlock(Block.Properties
                    .create(Material.SPONGE)
                    .harvestTool(ToolType.get("sword"))
                    .sound(SoundType.CLOTH)
                    .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    , 7
                    ,'L');
    public static Block lava_extra_large_sponge =
            new SpongeBlock(Block.Properties
                    .create(Material.SPONGE)
                    .harvestTool(ToolType.get("sword"))
                    .sound(SoundType.CLOTH)
                    .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    , 14
                    ,'E');
    //Sponge Blocks Items
    public static BlockItem large_spongeItem =
            new BlockItem(large_sponge,
                    new Item
                            .Properties()
                            .maxStackSize(64)
                            .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem extra_large_spongeItem =
            new BlockItem(extra_large_sponge,
                    new Item
                            .Properties()
                            .maxStackSize(64)
                            .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem lava_spongeItem =
            new BlockItem(lava_sponge,
                    new Item
                            .Properties()
                            .maxStackSize(64)
                            .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem lava_large_spongeItem =
            new BlockItem(lava_large_sponge,
                    new Item
                            .Properties()
                            .maxStackSize(64)
                            .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem lava_extra_large_spongeItem =
            new BlockItem(lava_extra_large_sponge,
                    new Item
                            .Properties()
                            .maxStackSize(64)
                            .group(ItemGroup.BUILDING_BLOCKS)
            );
    //Wet Sponge Blocks
    public static Block wet_large_sponge =
            new WetSpongeBlock(
                    Block.Properties
                            .create(Material.SPONGE)
                            .harvestTool(ToolType.get("sword"))
                            .sound(SoundType.CLOTH)
                            .hardnessAndResistance(0.5f, 0.5f)
                    ,false
                    ,'L');
    public static Block wet_extra_large_sponge =
            new WetSpongeBlock(
                    Block.Properties
                            .create(Material.SPONGE)
                            .harvestTool(ToolType.get("sword"))
                            .sound(SoundType.CLOTH)
                            .hardnessAndResistance(0.5f, 0.5f)
                    , false
                    ,'E');
    public static Block wet_lava_sponge =
            new WetSpongeBlock(
                    Block.Properties
                            .create(Material.SPONGE)
                            .harvestTool(ToolType.get("sword"))
                            .sound(SoundType.CLOTH)
                            .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    ,'N');
    public static Block wet_lava_large_sponge =
            new WetSpongeBlock(
                    Block.Properties
                            .create(Material.SPONGE)
                            .harvestTool(ToolType.get("sword"))
                            .sound(SoundType.CLOTH)
                            .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    ,'L');
    public static Block wet_lava_extra_large_sponge =
            new WetSpongeBlock(
                    Block.Properties
                            .create(Material.SPONGE)
                            .harvestTool(ToolType.get("sword"))
                            .sound(SoundType.CLOTH)
                            .hardnessAndResistance(0.5f, 0.5f)
                    , true
                    ,'E');
    //Wet Sponge Blocks Items
    public static BlockItem wet_large_spongeItem =
            new BlockItem(wet_large_sponge, new
                    Item
                            .Properties()
                    .maxStackSize(64)
                    .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem wet_extra_large_spongeItem =
            new BlockItem(wet_extra_large_sponge, new
                    Item
                            .Properties()
                    .maxStackSize(64)
                    .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem wet_lava_spongeItem =
            new BlockItem(wet_lava_sponge, new
                    Item
                            .Properties()
                    .maxStackSize(64)
                    .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem wet_lava_large_spongeItem =
            new BlockItem(wet_lava_large_sponge, new
                    Item
                            .Properties()
                    .maxStackSize(64)
                    .group(ItemGroup.BUILDING_BLOCKS)
            );
    public static BlockItem wet_lava_extra_large_spongeItem =
            new BlockItem(wet_lava_extra_large_sponge, new
                    Item
                            .Properties()
                    .maxStackSize(64)
                    .group(ItemGroup.BUILDING_BLOCKS)
            );

    @SubscribeEvent
    public static void registerBlocks(RegistryEvent.Register<Block> event) {
        event.getRegistry().registerAll(
                //Sponge Blocks Registry
                large_sponge.setRegistryName("large_sponge"),
                extra_large_sponge.setRegistryName("extra_large_sponge"),
                lava_sponge.setRegistryName("lava_sponge"),
                lava_large_sponge.setRegistryName("lava_large_sponge"),
                lava_extra_large_sponge.setRegistryName("lava_extra_large_sponge"),
                //Wet Sponge Blocks Registry
                wet_large_sponge.setRegistryName("wet_large_sponge"),
                wet_extra_large_sponge.setRegistryName("wet_extra_large_sponge"),
                wet_lava_sponge.setRegistryName("wet_lava_sponge"),
                wet_lava_large_sponge.setRegistryName("wet_lava_large_sponge"),
                wet_lava_extra_large_sponge.setRegistryName("wet_lava_extra_large_sponge")
        );
    }
    @SubscribeEvent
    public static void registerItem(RegistryEvent.Register<Item> event){
        event.getRegistry().registerAll(
                //Sponge Blocks Items Registry
                large_spongeItem.setRegistryName("large_sponge"),
                extra_large_spongeItem.setRegistryName("extra_large_sponge"),
                lava_spongeItem.setRegistryName("lava_sponge"),
                lava_large_spongeItem.setRegistryName("lava_large_sponge"),
                lava_extra_large_spongeItem.setRegistryName("lava_extra_large_sponge"),
                //Wet Sponge Blocks Items Registry
                wet_large_spongeItem.setRegistryName("wet_large_sponge"),
                wet_extra_large_spongeItem.setRegistryName("wet_extra_large_sponge"),
                wet_lava_spongeItem.setRegistryName("wet_lava_sponge"),
                wet_lava_large_spongeItem.setRegistryName("wet_lava_large_sponge"),
                wet_lava_extra_large_spongeItem.setRegistryName("wet_lava_extra_large_sponge")
        );
    }
}
